File: src/main/java/sudoku/Field.java
package sudoku;

import lombok.Data;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

@Data
public class Field {

    public final static List numbers = Arrays.asList(1, 2, 3, 4, 5, 6, 7, 8, 9);

    private FieldElement[][] fields;
    private Field prevField;
    private Point point;

    public static FieldElement[][] createField(int[][] field) {
        FieldElement[][] newField = new FieldElement[field.length][field[0].length];
        for (int i = 0; i < field.length; i++) {
            for (int j = 0; j < field[i].length; j++) {
                newField[i][j] = field[i][j] == 0 ? new FieldElement(0, new ArrayList<Integer>(numbers)) :
                        new FieldElement(field[i][j], Collections.<Integer>emptyList());
            }
        }
        return newField;
    }

    public FieldElement[][] copyFields() {
        FieldElement[][] array = new FieldElement[fields.length][fields[0].length];
        for (int i = 0; i < fields.length; i++) {
            for (int j = 0; j < fields[i].length; j++) {
                array[i][j] = new FieldElement(fields[i][j].getNumber(), new ArrayList(fields[i][j].getListOfNumber()));
            }
        }
        return array;
    }

    public Field pop() {
        Point point = prevField.getPoint();
        prevField.fields[point.getX()][point.getY()].getListOfNumber().remove((Object)point.getNumber());
        return prevField;
    }

}

File: src/main/java/sudoku/Field.kt
package sudoku

import java.util.*

data class Field(var fields: Array<Array<FieldElement>>, var prevField: Field?, var point: Point?) {

    fun copyFields(): Array<Array<FieldElement>> {
        val array = Array<Array<FieldElement>>(fields.size, { i -> Array<FieldElement>(fields.size, { FieldElement(0, mutableListOf()) })})
        for (i in fields.indices) {
            for (j in 0..fields[i].size - 1) {
                array[i][j] = FieldElement(fields[i][j].number, ArrayList(fields[i][j].listOfNumber))
            }
        }
        return array
    }

    fun pop(): Field {
        val point = prevField?.point
        prevField!!.fields[point!!.x][point!!.y].listOfNumber.remove(point.number as Any)
        return prevField!!
    }

    companion object {
        val numbers = Arrays.asList(1, 2, 3, 4, 5, 6, 7, 8, 9)

        fun createField(field: Array<Array<Int>>): Array<Array<FieldElement>> {
            val newField =  Array<Array<FieldElement>>(field.size, {i -> Array<FieldElement>(field.size, { FieldElement(0, mutableListOf()) })})
            for (i in field.indices) {
                for (j in 0..field[i].size - 1) {
                    newField[i][j] = if (field[i][j] == 0)
                        FieldElement(0, ArrayList<Int>(numbers))
                    else
                        FieldElement(field[i][j], mutableListOf())
                }
            }
            return newField
        }
    }

}